### ISSUE TITLE
Fody does not work in the dotnet cli

### ISSUE THREAD

**Issue by fealebenpae** [Fody/Fody#330]
I added Fody to a project created with `dotnet new` and set up my own `FodyWeavers.xml`. However, `dotnet build` fails with
```
/Users/yavorgeorgiev/.nuget/packages/fody/2.0.7/build/netstandard1.4/Fody.targets(53,5): error MSB4062: The "Fody.WeavingTask" task could not be loaded from the assembly /Users/yavorgeorgiev/.nuget/packages/fody/2.0.7/build/netstandard1.4/../../Fody.dll. Could not load file or assembly 'Microsoft.Build.Utilities.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified. [/Users/yavorgeorgiev/Projects/foo/foo.csproj]
/Users/yavorgeorgiev/.nuget/packages/fody/2.0.7/build/netstandard1.4/Fody.targets(53,5): error MSB4062:  Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask. [/Users/yavorgeorgiev/Projects/foo/foo.csproj]
    0 Warning(s)
    1 Error(s)
```

According to https://github.com/dotnet/cli/issues/5457 it seems that a way to solve this would be to provide two sets of assemblies - one built with the regular .NET Framework and another built with .NET Core, and switch between the two based on the value of `MSBuildRuntimeType`.

**Comment by SimonCropp** [comment]
so this would mean a breaking change in fody and every downstream weaver needing to move over to non-desktop runtime.

**Comment by fealebenpae** [comment]
The way I understand it it's only Fody itself that will have to provide both .NET Framework and .NET Core assemblies because of the MSBuild reference. Provided that weavers do not reference assemblies that don't exist in Core they should be fine.

Also, I think it's about providing both a .NET Framework *and* a .NET Core assembly, not moving over to .NET Core. `Fody.targets` would have to detect the runtime it's running on based on `MSBuildRuntimeType` and select the proper assembly to loads build tasks from.

**Comment by SimonCropp** [comment]
" Provided that weavers do not reference assemblies that don't exist in Core they should be fine."  

and that is exactly the problem, because they can and probably do. and asking all weavers to test in `dotnet build` is a bit too much. For example i know some weavers reference msbuild assemblies

**Comment by fealebenpae** [comment]
> some weavers reference msbuild assemblies

That's fine, I suppose, if they want to support `dotnet build` they will have to opt-in by shipping assemblies for both runtimes. If they don't care about .NET Core they will continue happily working on .NET Framework. But for weavers that want to support .NET Core and don't mind shipping two sets of assemblies Fody's lack of support for the .NET CLI is currently a blocker.

**Comment by tpluscode** [comment]
Wait, so any project that uses Fody will not build from the CLI? Sounds pretty important for anyone building in a CI pipeline ;)

**Comment by SimonCropp** [comment]
@tpluscode agreed. currently blocked by this https://github.com/jbevain/cecil/issues/406 ie i could not get a netstandard build of cecil

**Comment by SimonCropp** [comment]
but to be fair msbuild from the command line still works

**Comment by tpluscode** [comment]
Oh, I don't know why but I'd though that msbuild did not support the new project format. How silly

**Comment by SimonCropp** [comment]
@tpluscode i doubt many people are using dotnet build yet given it runs on .net core and is incompatible with many existing msbuild tasks

**Comment by fubar-coder** [comment]
We have the same problem here. The solution would be to switch from AppDomain to AssemblyLoadContext when Fody is run from the .NET Core CLI. You can differentiate between the two runtimes by querying `MSBuildRuntimeType` which is set to `Core` when using MSBuild running on .NET Core and is `Full` when running on the full .NET Framework.

To support putting `FodyWeavers.xml` into the project folder, you should place it into the [`contentFiles`](https://docs.microsoft.com/de-de/nuget/schema/nuspec#including-content-files) folder instead of the content folder.

**Comment by BrunoJuchli** [comment]
@fubar-coder 
`contentFiles` is for `packageReference` style nuget-references and the behavior is different from the "old" `content` folder (that is still being used for the old csproj / nuget references). With `contentFiles` the files are supposed to be immutable as they are replaced on every restore. However, an immutable `FodyWeavers.xml` would be pretty useless...
As i read it there's currently no feature-equivalent replacement for placing files in `content` folder, the recommended workarounds are to use `init.ps1` and/or adding an extra `*.targets` file, but these will be run much more often than "on install" or "on upgrade". Hence it's tricky to get right.

**Comment by WallaceKelly** [comment]
> i doubt many people are using dotnet build yet 

@simoncropp, that my have been true back in June, but not any longer. :-)

**Comment by lthompson-cartasite** [comment]
Anyone using Amazon's Lambda microservice tech is using dotnet build whether they mean to or not. It is the only build chain offered, and this limitation forced us to step away from C# and use node.js instead despite our strong history with C# in-house.

**Comment by TylerBrinkley** [comment]
Blocked on setting up CI for [Enums.NET](https://github.com/TylerBrinkley/Enums.NET) because of this.

**Comment by per-samuelsson** [comment]
Any news on this? We are considering testing Fody out for one of our projects and this is unfortunately a showstopper for us.

Based on my understanding, I don't see this thing would affect weaver implementations, but rather is isolated to the plumbing supported by the Fody engine, no? If so, it shouldn't be a major effort to address.

**Comment by MihaMarkic** [comment]
I'd really like to see it working as well.
While MSBuild can be used, there is another collateral: nunit tests won't run because they run only using _dotnet test_ which doesn't work with fody due to this problem.

**Comment by MihaMarkic** [comment]
OTOH one can use a combination of MSBuild and **dotnet test --no-build**. So, not really that important anymore to me since that's a simple workaround.

**Comment by StephenDrewLDS** [comment]
Is this closed as done?  Build log below using Costura...

> [build] C:\Users\xxx\.nuget\packages\fody\2.0.0\build\netstandard1.4\Fody.targets(53,5): error MSB4062: The "Fody.WeavingTask" task could not be loaded from the assembly C:\Users\xxx\.nuget\packages\fody\2.0.0\build\netstandard1.4\..\..\\Fody.dll. Could not load file or assembly 'Microsoft.Build.Utilities.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
> 
> C:\Users\xxx\.nuget\packages\fody\2.0.0\build\dotnet\Fody.targets(53,5): error MSB4062: The "Fody.WeavingTask" task could not be loaded from the assembly C:\Users\xxx\.nuget\packages\fody\2.0.0\build\dotnet\..\..\\Fody.dll. Could not load file or assembly 'Microsoft.Build.Utilities.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.

> [17:17:46][build] Process exited with code 1

**Comment by drauch** [comment]
Is this considered fixed or has it been closed otherwise? We need to deploy our applications using "dotnet publish", which doesn't work atm...

**Comment by SimonCropp** [comment]
As per the milestone, this was fixed in 2.3.18. If you are still having issues i suggest u raise a new bug

**Comment by drauch** [comment]
As we use MSBuild publishing instead of dotnet publish now, I don't have time to investigate further. Will raise a bug if I ever run into it again. Thanks for the quick reply though!
